### ISSUE TITLE
Import spreadsheets without a raw header results in an error

### ISSUE THREAD

**Issue by adrian-lara** [SEED-platform/seed#2374]
### Expected Behavior
TBD: it's an open question regarding how this should be handled in SEED, if at all.

An easy approach is to _not_ handle it, and prevent users from continuing with an error message on the mapping page indicating that the file must have a header.

If we do want to be able to handle files with missing headers, more brainstorming is needed. 
For handling files like this:
- ~One approach we can take is to just take the mapped column name and use it as the raw column name. Downstream impacts will need to be considered further, but my initial thinking is this wouldn't cause any breaking issues. Further consideration might also be needed for mapping profiles - i.e consider what happens when a user tries to save a preset with a missing raw header.~ This wouldn't work as the raw PropertyState objects holding everything in extra_data would not have keys for these values.
- Could there be filler header names we could use in these cases?

### Actual Behavior
When importing a file, if one of the incoming raw headers is blank,
![Screen Shot 2020-08-04 at 11 19 03 AM](https://user-images.githubusercontent.com/30608004/90664320-0c2b4700-e208-11ea-80f8-46f7b8d734be.png)

SEED will throw an error.
```
DoesNotExist: Column matching query does not exist.
  File "django/db/models/query.py", line 464, in get_or_create
    return self.get(**lookup), False
  File "django/db/models/query.py", line 380, in get
    self.model._meta.object_name
 
ValidationError: {'column_name': ['This field cannot be blank.']}
(19 additional frame(s) were not displayed)
...
  File "seed/lib/superperms/orgs/decorators.py", line 244, in _wrapped
    return fn(self, request, *args, **kwargs)
  File "seed/data_importer/views.py", line 1043, in save_column_mappings
    result = Column.create_mappings(mappings, organization, request.user, import_file.id)
  File "seed/models/columns.py", line 814, in create_mappings
    mappings = Column._column_fields_to_columns(mappings, organization)
  File "seed/models/columns.py", line 982, in _column_fields_to_columns
    is_extra_data=False  # data from header rows in the files are NEVER extra data
  File "seed/models/columns.py", line 1397, in validate_model
    instance.full_clean()
```

### Instance Information
all

**Comment by RDmitchell** [comment]
@adrian-lara -- I think we should put up a message to the user that they can't have blank headers, and that they need to add a header and reimport the file.

Somehow I thought this error checking already existed, but I guess not.

**Comment by adrian-lara** [comment]
> @adrian-lara -- I think we should put up a message to the user that they can't have blank headers, and that they need to add a header and reimport the file.
> 
> Somehow I thought this error checking already existed, but I guess not.

I think this would be a great quick fix.

Do you think that users might find it useful if SEED could actually handle/import these types of files successfully?

**Comment by RDmitchell** [comment]
@adrian-lara -- I think you would have to allow the user to specify what they wanted the column name(s) to be for each blank header, which might be a bit of work, if you want SEED to handle it without having the user reimport a corrected file.

**Comment by adrian-lara** [comment]
My initial thought was that users can sort of "specify" that now by mapping to SEED columns, and maybe we could use those mapped as our "stand in" for any blank headers. Generally speaking though, you're right in that this would require a bit more work than the approach of providing an error message and preventing a full import.

I guess I'm mostly just wondering if the ability to handle empty raw headers is worth more consideration. Are there prospective users or current users that could benefit from this?

**Comment by RDmitchell** [comment]
I'm sure that some users would benefit from SEED handling this within the program, I'm just not sure how high a priority it should have right now. I guess that might be something for @nllong to decide.

**Comment by adrian-lara** [comment]
Per separate discussions, addressing this in some way will help tighten up SEED in general, which is something we'd definitely like to do at the moment. Given that, the current plan is to address this for the upcoming release this month.

**Comment by macintoshpie** [comment]
I see these choices as our options
1) Just give a warning to user and prevent import. This is easy to implement, but could be frustrating for users.

2) Auto generate header names for any that are missing at early stage of import (possibly within tasks.py cache_first_rows or individual readers/parser). This is easy to implement, and convenient to users, though maybe confusing when they see a generated header.

3) Allow users to create custom names for fields. This will require more work on the backend (updating the import_file view, and probably tweaks to the import file model), and is convenient for users and probably the least confusing.

I'd argue that 2 is our best option. Raw headers are only used when mapping the file to give suggested mappings and for the user to know what data they're referencing on the mapping page. After mapping raw headers don't matter to the user at all, so they probably wouldn't care if we warn them and generate one for them. This would be a pretty trivial change to the backend, and minimal effort on the front end. Option 3 is nice but I think the complexity it would introduce is not worth it.

Proposed User Flow:
1. user selects a file missing header on data upload modal
2. user gets the "successful upload" modal, _but_ we include a warning that there were missing columns below the `example-pm-properties.xlsx has been uploaded to abcd.` success message
```
Warning: your file was missing one or more headers which have been replaced with auto-generated names.
This will not affect your data import.
If you prefer to use your own headers please select "Cancel upload", otherwise you can continue to data mapping.
```
3. if user clicks cancel upload, front end makes call to delete the import file created and the modal is dismissed. Otherwise the user can click continue to data mapping like normal, and is redirected to mapping page
4. on the mapping page, the user will see `SEED Generated Header 000` wherever there were headers missing. They will have incremental numbers.
5. import continues as normal

**Comment by RDmitchell** [comment]
@macintoshpie / @adrian-lara -- I think option 2 as described makes sense.

**Comment by adrian-lara** [comment]
I agree - it's a good solution and sounds reasonable to implement!

**Comment by macintoshpie** [comment]
Note that the proposed flow has been slightly modified s.t. the modal is shown on the mapping page rather than in the data-upload modal

**Comment by RDmitchell** [comment]
instance: dev1
SHA: 38fb067
Org: LBNL Test 200
Cycle: 2015 Compliance Cycle

@macintoshpie / @adrian-lara -- This doesn't appear to be working. Maybe I am misunderstanding the test I am supposed to do.

I have a file with one field that has no column header
![image](https://user-images.githubusercontent.com/6314950/94327425-d6445580-ff5f-11ea-9d7c-c7649d67dfdb.png)

Program seems to be hanging on the initial import of this file
![image](https://user-images.githubusercontent.com/6314950/94327447-fa079b80-ff5f-11ea-8da8-8131e0790533.png)

**Comment by adrian-lara** [comment]
@macintoshpie given this comment: 
>I did not implement this functionality for excel spreadsheets b/c the way the parser works makes it pretty unlikely for there to be an empty header. Basically, the header of a spreadsheet is determined by finding the first row that has no empty cells.
https://github.com/SEED-platform/seed/pull/2391#issuecomment-691251089

At the very least, I wouldn't have expected the upload to hang. What do you think?

**Comment by RDmitchell** [comment]
@adrian-lara / @macintoshpie -- my test case is one where there is one blank field in the input file, rather than no header at all. 

I guess I am not really understanding the use case of this fix.

**Comment by macintoshpie** [comment]
@RDmitchell Thanks for testing this!

The use case of the issue, from my understanding, is a missing _header_ in an imported file. After doing a bit of research I realized the use case was actually even more specific, importing a _csv_ with a missing header. We can't easily handle the case for spreadsheets (.xslx files) b/c of the format of some spreadsheet files. I think the example you gave of an xlsx file missing a header is valuable though, and we need to gracefully fail, I'll look into what's happening there.

As an example of why it's not easy to auto-generate missing headers for spreadsheets, [this PM properties file](https://github.com/SEED-platform/seed/blob/develop/seed/data_importer/tests/data/example-pm-properties.xlsx) has multiple "partial" rows before we reach the actual header row. The way our "algorithm" finds the header is to **find the first row with no empty cells**. Unless we change this approach (which I don't think we want to, there could be larger implications), then we can't replace empty header cells because we'd never find a missing header.

I think the use case you're describing, "one blank field in the input file" is already covered, that property/tax lot just won't have a value for that column, and you might get a warning about it from the data quality checker.

**Comment by RDmitchell** [comment]
@macintoshpie -- the "one blank field in the input file" that I am referring to is a blank field name in the header, assuming that I understand your definition of header. 

I assume by "header" you mean the row in the CSV (or spreadsheet) that contains the column names for the data that follows (see my screen shot at the beginning of this issue).

Looking at @adrian-lara original statement of the problem, his screenshot shows an individual field in the mapping screen that didn't have a field name in the "header" and is therefore blank in the mapping, which is the case I am talking about. 

I guess not being able to come up automatically with a default field name is OK, but the program needs to make sure to handle that case gracefully, as you said.

**Comment by RDmitchell** [comment]
@macintoshpie / @adrian-lara -- so what is the status of this issue? I guess I'm not quite sure.

**Comment by macintoshpie** [comment]
@RDmitchell This issue should be completed after #2391 is approved and merged.

This new PR will just fail if a user tries to import an invalid xlsx file (ie it will not generate headers, see my comment above about why we cannot do this for spreadsheet/xlsx files but we can for CSVs). This should address the issue you had above importing an XLSX file with a missing header that just hung on the import dialog.

**Comment by RDmitchell** [comment]
@macintoshpie -- is it on dev1?

**Comment by macintoshpie** [comment]
Oh sorry! I forgot we have to manually push those changes, it is not on dev1. I guess we will have to wait until the next round of changes are pushed to the server.

**Comment by RDmitchell** [comment]
@macintoshpie -- is it on dev1 yet? Just checking on the last 2 issues that I'm supposed to test for this github project. Thx

**Comment by RDmitchell** [comment]
@adrian-lara -- I can't test this until the code is on dev1, so if someone can let me know when that happens, then I will test. Thx

**Comment by adrian-lara** [comment]
@RDmitchell I think you were on dev1 already, but just in case, it's up as of this writing 😄

**Comment by RDmitchell** [comment]
Instance: dev1
SHA: 
Org: LBNL 200 Test
File: portfolio-manager-sample-missing-headers.csv

Looking at PR https://github.com/SEED-platform/seed/pull/2391
where the Manual Test instructions were listed, doesn't seem like the program is working exactly as stated there

Message when I import that file
![image](https://user-images.githubusercontent.com/6314950/98489091-5a396080-21e1-11eb-81f3-27bbe8af4132.png)
Doesn't seem like auto-generated names were created, but I don't think that really matters. The program doesn't crash and there are several ways the user can map the fields. 
![image](https://user-images.githubusercontent.com/6314950/98489144-9cfb3880-21e1-11eb-9eee-c89d40a8fe08.png)

**Comment by RDmitchell** [comment]
instance: dev1
SHA: 
Org: LBNL Test 200
File: portfolio-manager-sample-missing-headers.csv

Here is the message the program put up, which doesn't seem to be quite the same as discussed in #2391
![image](https://user-images.githubusercontent.com/6314950/98489431-d41e1980-21e2-11eb-9bd6-89694bdc9cb5.png)
Program didn't make any field name suggestions, but I think that's ok, because the user has lots of options at this point since at least the program didn't hang.
![image](https://user-images.githubusercontent.com/6314950/98489467-e8faad00-21e2-11eb-9041-7a262e0b8386.png)

**Comment by RDmitchell** [comment]
Instance: dev1
SHA; e2078bfbe
Org: LBNL Test 200
File: portfolio-manager-sample-missing-headers.csv except saved as XLSX file

Program seems to be hanging. 
![image](https://user-images.githubusercontent.com/6314950/98489509-22331d00-21e3-11eb-92ff-2ebd313acfcf.png)

**Comment by RDmitchell** [comment]
@adrian-lara / @macintoshpie 

Not sure what to do with this. We could just close it even though the program still seems to hang on importing XLSX files.

Feel free to close it if you want to.

**Comment by RDmitchell** [comment]
@adrian-lara / @macintoshpie

I put the two files I imported into the issue folder
https://drive.google.com/drive/folders/1ovXJxMSEj7o6D3a6lswvQCoY7HoZiO3z?usp=sharing

I got the original CSV file from the repo test folder.

**Comment by macintoshpie** [comment]
Thanks for testing Robin. I'll spend a minute looking into the issue with the hanging xlsx file, I was able to replicate your result on dev1 as well. For the CSV, the autogenerated name is just in the Data File Header column. It's up to the user to provide a mapping to a SEED column.

**Comment by RDmitchell** [comment]
Got it ! Makes sense. Thx

**Comment by nllong** [comment]
Sounds like we should move this to the Dec project board. I'm dumping it in the bottom of the dec 2020 todo.

**Comment by macintoshpie** [comment]
@RDmitchell I'm not able to reproduce the hanging error when importing the excel file now on dev1 (2.9.0.e84e9f7bb) or current develop (6fb7d1bd4721682826c68665a6f53aacc2d448b1). Could you test this again and let me know if you're seeing the issue still?

**Comment by RDmitchell** [comment]
Instance: LBNL staging
SHA: e84e9f7bb
Org: LBNL Test 200
Cycle: Import No Header

This worked great !!

SEED put in auto-generated header names (after the note saying it would do that), I mapped them to other fields, and the import when smoothly.
![image](https://user-images.githubusercontent.com/6314950/106675065-4051b480-6569-11eb-9852-f389725e7d2d.png)

I'm closing !
